test.rb

```diff
@@ -0,0 +1,251 @@
+            (1...@hash_array.size).each do |num|
+                if  num < @hash_array.size && @hash_array[num][group_by_symbol] == @hash_array[num - 1][group_by_symbol]
+
+                    math_total += @hash_array[num][:maths]
+                    physics_total += @hash_array[num][:physics]
+                    chemistry_total +=  @hash_array[num][:chemistry]
+
+                else
+                    # puts "#{num}  #{count} "
+                    no_students = num - count
+                    math_average = math_total / no_students
+                    physics_average = physics_total / no_students
+                    chemistry_average = chemistry_total / no_students
+                    # puts "Phase #{num} - #{math_total} #{physics_total} #{chemistry_total} "
+                    htemp_groupby = { group_by_symbol => @hash_array[num - 1][group_by_symbol], :maths => math_average, :physics => physics_average, :chemistry => chemistry_average }
+                    @hash_array_group_by.push(htemp_groupby)
+
+                    math_total = @hash_array[num][:maths]
+                    physics_total = @hash_array[num][:physics]
+                    chemistry_total = @hash_array[num][:chemistry]
+                    count = num
+
+                    # puts "Phase #{num} - #{math_average} #{physics_average} #{chemistry_average} "
+                end
+            end
+            no_students = @hash_array.size - count
+            math_average = math_total / no_students
+            physics_average = physics_total / no_students
+            chemistry_average = chemistry_total / no_students
+            htemp_groupby = { group_by_symbol => @hash_array[@hash_array.size - 1][group_by_symbol], :maths => math_average, :physics => physics_average, :chemistry => chemistry_average }
+            @hash_array_group_by.push(htemp_groupby)
+
+        else
+            puts 'Enter Some valid input'
+      end
+  end
+
+    # Sort_by method
+    def student_sort_data(sort_by)
+        if sort_by.casecmp('maths').zero? || sort_by.casecmp('physics').zero? || sort_by.casecmp('chemistry').zero?
+            sort_by_symbol = sort_by.downcase.to_sym
+            @hash_array_group_by = @hash_array_group_by.sort_by { |hsh| hsh[sort_by_symbol] }.reverse
+        else
+            puts 'Enter Some valid input'
+        end
+    end
+
+    def print_student_array
+        @hash_array_group_by.each do |element|
+            puts element
+        end
+    end
+
+    def display_student_data(display_fields, group_by, show_total)
+        display_fields = display_fields.split(',')
+        group_by = group_by.upcase
+        total_physics = 0
+        total_maths = 0
+        total_chemistry = 0
+        display_fields.map!(&:downcase)
+        print_student_array
+        print " |   #{group_by}   "
+        display_fields.each do |field|
+            print " |#{field}       "
+        end
+        print "\n"
+
+        @hash_array_group_by.each do |element|
```

Comment:
why are you calculating total multiple times? students_group_data and display_student_data